cmd/cstor-pool-mgmt/controller/backup-controller/new_backup_controller.go

```diff
@@ -0,0 +1,246 @@
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: backupControllerName})
+
+	controller := &BackupController{
+		kubeclientset: kubeclientset,
+		clientset:     clientset,
+		BackupSynced:  BackupInformer.Informer().HasSynced,
+		workqueue:     workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "BackupCStor"),
+		recorder:      recorder,
+	}
+
+	glog.Info("Setting up event handlers for backup")
+
+	// Clean any pending backup for this cstor pool
+	controller.cleanupOldBackup(clientset)
+
+	// Instantiating QueueLoad before entering workqueue.
+	q := common.QueueLoad{}
+
+	// Set up an event handler for when BackupCStor resources change.
+	BackupInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			// Note: AddFunc is called when a new object comes into etcd
+			// Note : In case controller reboots and existing object in etcd can cause delivery of
+			// add event when the controller comes again. Be careful in this part and handle accordingly.
+			bkp := obj.(*apis.BackupCStor)
+
+			if !IsRightCStorPoolMgmt(bkp) {
+				return
+			}
+			controller.handleBKPAddEvent(bkp, &q)
+		},
+		UpdateFunc: func(oldVar, newVar interface{}) {
+			// Note : UpdateFunc is called in following three cases:
+			// 1. When object is updated/patched i.e. Resource version of object changes.
+			// 2. When object is deleted i.e. the deletion timestamp of object is set.
+			// 3. After every resync interval.
+			newbkp := newVar.(*apis.BackupCStor)
+			oldbkp := oldVar.(*apis.BackupCStor)
+
+			if !IsRightCStorPoolMgmt(newbkp) {
+				return
+			}
+
+			controller.handleBKPUpdateEvent(oldbkp, newbkp, &q)
+		},
+		DeleteFunc: func(obj interface{}) {
+			bkp := obj.(*apis.BackupCStor)
+			if !IsRightCStorPoolMgmt(bkp) {
+				return
+			}
+			glog.Infof("BackupCStor Resource delete event: %v, %v", bkp.ObjectMeta.Name, string(bkp.ObjectMeta.UID))
+		},
+	})
+	return controller
+}
+
+// enqueueBackupCStor takes a BackupCStor resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than BackupCStor.
+func (c *BackupController) enqueueBackupCStor(obj *apis.BackupCStor, q common.QueueLoad) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	q.Key = key
+	c.workqueue.AddRateLimited(q)
+}
+
+// cleanupOldBackup set fail status to old pending backup
+func (c *BackupController) cleanupOldBackup(clientset clientset.Interface) {
+	bkplabel := "cstorpool.openebs.io/uid=" + os.Getenv(string(common.OpenEBSIOCStorID))
+	bkplistop := metav1.ListOptions{
+		LabelSelector: bkplabel,
+	}
+	bkplist, err := clientset.OpenebsV1alpha1().BackupCStors(metav1.NamespaceAll).List(bkplistop)
+	if err != nil {
+		return
```

Comment:
```go
return errors.Wrapf(err, "failed to cleanup old backups: list options {%s}", bkplistop)
```

Reply:
What happens when we fail to cleanup.
I guess they remain in this state forever till application is restarted?

Reply:
yes